Add cookbook - Test your code with java 8 (#179)
VirtusLab/scala-cli: website/docs/cookbooks/scala-jvm.md

---
title: Test your code with different Java versions
sidebar_position: 4
---

> You can use Scala CLI to test your code compatibility with various versions of `java` (no manual installation of java is required).It automatically downloads the specified version of java.

Reviewer: ```suggestion
You can use Scala CLI to test your code compatibility with various versions of `java` (no manual installation of jdk or sdk is required). It automatically downloads the specified version of java.
```

> The following snippet used the new method `Files.writeString` from `java` 11. 

Reviewer 2: ```suggestion
The following snippet uses the new method `Files.writeString` from Java 11. 
```

```scala name:Main.scala
import java.nio.file.Files
import java.nio.file.Paths

object Main extends App {

>     val dest = Files.createTempFile(
>        Paths.get("").toAbsolutePath(),
>        "scala-cli",
>        ".txt"
>     ) 
>     val filePath = Files.writeString(dest, "Hello from ScalaCli")
>     val fileContent: String = Files.readString(filePath)
>   }

Reviewer: ```suggestion
  val dest = Files.createTempFile(
    Paths.get("").toAbsolutePath(),
      "scala-cli",
      ".txt"
  ) 
  val filePath = Files.writeString(dest, "Hello from ScalaCli")
  val fileContent: String = Files.readString(filePath)
}
```

 ```

> Passing `--jvm` to the `scala-cli` command and run your application with the specified java version.

Reviewer 2: ```suggestion
Pass `--jvm` to the `scala-cli` command to run your application with the specified java version.
```

```scala-cli 
scala-cli Main.scala --jvm 11
```

<!-- Expected:

> -->

Reviewer: No expectations?

> To test your application with java 8, change the value of `--jvm` parameter.

Reviewer 2: ```suggestion
To test your application with Java 8, change the value of `--jvm` parameter.
```

---

> title: Test your code with java 8 

Reviewer: ```suggestion
title: Test your code with different Java veresions
```

sidebar_position: 4
---

You can use Scala CLI to test your code compatibility with various versions of `java` (no manual installation of java is required).It automatically downloads the specified version of java.

The following snippet used the new method `Files.writeString` from `java` 11. 

```scala name:Main.scala
import java.nio.file.Files
import java.nio.file.Paths

object Main extends App {

>     val filePath = Files.writeString(
>       Files.createTempFile(
>         Paths.get("").toAbsolutePath(),
>         "scala-cli",
>         ".txt"
>       ),
>       "Hello from ScalaCli"
>     )

Reviewer: ```suggestion
    val dest = Files.createTempFile(
       Paths.get("").toAbsolutePath(),
       "scala-cli",
       ".txt"
    ) 
    val filePath = Files.writeString(dest, "Hello from ScalaCli")
```

    val fileContent: String = Files.readString(filePath)
  }
 ```

Passing `--jvm` to the `scala-cli` command and run your application with the specified java version.

```scala-cli 

> scala-cli Main.scala --jvm adopt:11

Reviewer: @alexarchambault would it be possible to specify just `--jvm 11` that will default to adopt-11?

Author: Yes, it is possible to specify just `--jvm 11` or `--jvm 8` etc. AdoptOpenJDK is used by default.

Here is more details about resolving jvm - [coursier](https://get-coursier.io/docs/cli-java#short-jvm-names)

Reviewer 2: We could default to Java 11. The default JVM id is currently defined [here](https://github.com/VirtusLab/scala-cli/blob/46416b11d3a15ad782dd3f371c02d87e7bb4f749/modules/build/src/main/scala/scala/build/internal/OsLibc.scala#L60-L62).

```

<!-- Expected:
-->

To test your application with java 8, change the value of `--jvm` parameter.
```bash
scala-cli Main.scala  --jvm adopt:8
# In this case, it raises an error because the `Files.createTempFile` method is not available in java 8
#

> # Exception in thread main: java.lang.Exception: java.lang.NoSuchMethodError: java.nio.file.Files.writeString(Ljava/nio/file/Path;Ljava/lang/CharSequence;[Ljava/nio/file/OpenOption;)Ljava/nio/file/Path;

Reviewer: @alexarchambault isn't it bad behaviour? Should code simply does not compile?

Author: I specially prepared a code snippet that shouldn't compile in jvm 8.

Reviewer 2: JVM switching only works well at runtime, and less so for compiling. When compiling, the JVM of Bloop is actually used.

Reviewer: but there are flags (`--target` if I am not mistaken) to force scala compiler to check if this will compile against given JVM. @prolativ can help here
